### ISSUE TITLE
Incorrect constraint violation error using `keyof` with generic

### ISSUE THREAD

**Issue by RyanCavanaugh** [Microsoft/TypeScript#12311]
**TypeScript Version:**  master

**Code**
```ts
interface HTMLKind {
    "a": HTMLAnchorElement;
    "div": HTMLDivElement;
    "span": HTMLSpanElement;
    "input": HTMLInputElement;
    // ... dozens more ...
    [s: string]: HTMLElement;
}

// Incorrect error on the return type here
declare function getElementsByTagName<T extends keyof HTMLKind>(k: T): NodeListOf<HTMLKind[T]>;
```

**Expected behavior:**
No error

**Actual behavior:**
> `Type 'HTMLKind[T]' does not satisfy the constraint 'Node'`.

**Comment by mhegazy** [comment]
This is blocking the change to lib.d.ts to reduce symbol instantiations at the moment, see https://github.com/Microsoft/TSJS-lib-generator/pull/166
